File: client-ts/src/resources.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 AimpathyMinds

/**
 * Fluent resource accessors for the YAAgents Agentic REST Profile v0.1.
 *
 * Usage:
 * ```ts
 * const result = await client.campaigns.byId("c1").optimizations().create({ goal: "ctr" });
 * const asset  = await client.campaigns.byId("c1").assets().generate({ format: "banner" });
 * ```
 */

import { parseResponse } from "./mapper.js";
import type { AgenticResult } from "./result.js";
import type { RequestOptions } from "./types.js";
// type-only import breaks the client ↔ resources circular reference at runtime
import type { YaAgentsClient } from "./client.js";

// ---------------------------------------------------------------------------
// OptimizationsResource
// ---------------------------------------------------------------------------

/** Sub-resource for campaign optimizations. */
export class OptimizationsResource {
  readonly #client: YaAgentsClient;
  readonly #campaignId: string;

  constructor(client: YaAgentsClient, campaignId: string) {
    this.#client = client;
    this.#campaignId = campaignId;
  }

  /**
   * POST /campaigns/{id}/optimizations — create a new optimization run.
   *
   * @param body - Request payload (framework-specific; see your gateway's OpenAPI spec).
   * @param options - Per-call options such as a custom `correlationId`.
   * @returns `AgenticResult<T>` — never throws; switch on `result.type`.
   *          Use `client.strict()` for exception-style handling.
   */
  async create<T = Record<string, unknown>>(
    body: Record<string, unknown>,
    options?: RequestOptions,
  ): Promise<AgenticResult<T>> {
    const response = await this.#client._request(
      "POST",
      `/campaigns/${this.#campaignId}/optimizations`,
      body,
      options,
    );
    return parseResponse<T>(response);
  }
}

// ---------------------------------------------------------------------------
// AssetsResource
// ---------------------------------------------------------------------------

/** Sub-resource for campaign asset generation. */
export class AssetsResource {
  readonly #client: YaAgentsClient;
  readonly #campaignId: string;

  constructor(client: YaAgentsClient, campaignId: string) {
    this.#client = client;
    this.#campaignId = campaignId;
  }

  /**
   * POST /campaigns/{id}/assets:generate — trigger asset generation.
   *
   * @param body - Request payload.
   * @param options - Per-call options such as a custom `correlationId`.
   * @returns `AgenticResult<T>` — never throws; switch on `result.type`.
   *          Use `client.strict()` for exception-style handling.
   */
  async generate<T = Record<string, unknown>>(
    body: Record<string, unknown>,
    options?: RequestOptions,
  ): Promise<AgenticResult<T>> {
    const response = await this.#client._request(
      "POST",
      `/campaigns/${this.#campaignId}/assets:generate`,
      body,
      options,
    );
    return parseResponse<T>(response);
  }
}

// ---------------------------------------------------------------------------
// CampaignResource
// ---------------------------------------------------------------------------

/** Fluent accessor for a single campaign and its sub-resources. */
export class CampaignResource {
  readonly #client: YaAgentsClient;
  /** Campaign identifier used in every sub-resource path. */
  readonly campaignId: string;

  constructor(client: YaAgentsClient, campaignId: string) {
    this.#client = client;
    this.campaignId = campaignId;
  }

  /** Return an {@link OptimizationsResource} scoped to this campaign. */
  optimizations(): OptimizationsResource {
    return new OptimizationsResource(this.#client, this.campaignId);
  }

  /** Return an {@link AssetsResource} scoped to this campaign. */
  assets(): AssetsResource {
    return new AssetsResource(this.#client, this.campaignId);
  }
}

// ---------------------------------------------------------------------------
// CampaignsAccessor
// ---------------------------------------------------------------------------

/**
 * Entry point for the campaigns resource collection.
 * Obtained via `client.campaigns`.
 */
export class CampaignsAccessor {
  readonly #client: YaAgentsClient;

  constructor(client: YaAgentsClient) {
    this.#client = client;
  }

  /**
   * Scope subsequent calls to a specific campaign.
   *
   * @param id - Campaign identifier.
   * @returns A {@link CampaignResource} fluent accessor.
   *
   * @example
   * ```ts
   * await client.campaigns.byId("c1").optimizations().create({ goal: "ctr" });
   * ```
   */
  byId(id: string): CampaignResource {
    return new CampaignResource(this.#client, id);
  }
}

File: client-python/src/yaagents_client/_resources.py
"""Fluent resource accessors for YAAgents Agentic REST Profile v0.1."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

import httpx

if TYPE_CHECKING:
    from ._client import YaAgentsClient

__all__ = ["CampaignResource"]


class OptimizationsResource:
    """Optimizations sub-resource scoped to one campaign."""

    def __init__(self, client: YaAgentsClient, campaign_id: str) -> None:
        self._client = client
        self._campaign_id = campaign_id

    def create(
        self,
        body: dict[str, Any],
        *,
        correlation_id: str | None = None,
    ) -> httpx.Response:
        """POST /campaigns/{id}/optimizations — create a new optimization run."""
        return self._client._request(
            "POST",
            f"/campaigns/{self._campaign_id}/optimizations",
            body,
            correlation_id=correlation_id,
        )


class AssetsResource:
    """Assets sub-resource scoped to one campaign."""

    def __init__(self, client: YaAgentsClient, campaign_id: str) -> None:
        self._client = client
        self._campaign_id = campaign_id

    def generate(
        self,
        body: dict[str, Any],
        *,
        correlation_id: str | None = None,
    ) -> httpx.Response:
        """POST /campaigns/{id}/assets:generate — trigger asset generation."""
        return self._client._request(
            "POST",
            f"/campaigns/{self._campaign_id}/assets:generate",
            body,
            correlation_id=correlation_id,
        )


class CampaignResource:
    """Fluent resource accessor for a single campaign."""

    def __init__(self, client: YaAgentsClient, campaign_id: str) -> None:
        self._client = client
        self._campaign_id = campaign_id
        self.optimizations: OptimizationsResource = OptimizationsResource(
            client, campaign_id
        )
        self.assets: AssetsResource = AssetsResource(client, campaign_id)
